beacon-chain/core/altair/sync_committee.go

```
@@ -0,0 +1,163 @@
+	}, nil
+}
+
+// NextSyncCommitteeIndices returns the next sync committee indices for a given state.
+//
+// Spec code:
+// def get_next_sync_committee_indices(state: BeaconState) -> Sequence[ValidatorIndex]:
+//    """
+//    Return the sync committee indices, with possible duplicates, for the next sync committee.
+//    """
+//    epoch = Epoch(get_current_epoch(state) + 1)
+//
+//    MAX_RANDOM_BYTE = 2**8 - 1
+//    active_validator_indices = get_active_validator_indices(state, epoch)
+//    active_validator_count = uint64(len(active_validator_indices))
+//    seed = get_seed(state, epoch, DOMAIN_SYNC_COMMITTEE)
+//    i = 0
+//    sync_committee_indices: List[ValidatorIndex] = []
+//    while len(sync_committee_indices) < SYNC_COMMITTEE_SIZE:
+//        shuffled_index = compute_shuffled_index(uint64(i % active_validator_count), active_validator_count, seed)
+//        candidate_index = active_validator_indices[shuffled_index]
+//        random_byte = hash(seed + uint_to_bytes(uint64(i // 32)))[i % 32]
+//        effective_balance = state.validators[candidate_index].effective_balance
+//        if effective_balance * MAX_RANDOM_BYTE >= MAX_EFFECTIVE_BALANCE * random_byte:
+//            sync_committee_indices.append(candidate_index)
+//        i += 1
+//    return sync_committee_indices
+func NextSyncCommitteeIndices(state state.BeaconStateAltair) ([]types.ValidatorIndex, error) {
+	epoch := helpers.NextEpoch(state)
+	indices, err := helpers.ActiveValidatorIndices(state, epoch)
+	if err != nil {
+		return nil, err
+	}
+	count := uint64(len(indices))
+	seed, err := helpers.Seed(state, epoch, params.BeaconConfig().DomainSyncCommittee)
+	if err != nil {
+		return nil, err
+	}
+	i := types.ValidatorIndex(0)
+	cIndices := make([]types.ValidatorIndex, 0, params.BeaconConfig().SyncCommitteeSize)
+	hashFunc := hashutil.CustomSHA256Hasher()
+	maxRandomByte := uint64(1<<8 - 1)
+	for uint64(len(cIndices)) < params.BeaconConfig().SyncCommitteeSize {
```

Is this loop expensive? It may be worth having a context check to fail faster if the deadline has been exceeded.